feat(blog): How to remove outdated plugins while using docker (#6460)
jenkins-infra/jenkins.io: content/blog/2023/06/19/2023-06-19-remove-outdated-plugins-while-using-docker.adoc

### Removing deprecated plugins in the UI

Follow these steps to remove the deprecated plugin:

1. Click on *Plugins*, then select *Installed plugins*.
2. In the search box, enter the name of the deprecated plugin (in this case, it's `popper`).

image::/images/post-images/2023/06/19/2023-06-19-remove-outdated-plugins-while-using-docker/popper-can-t-be-uninstalled.png[width=839,search for the deprecated plugin]

Unfortunately, we encounter an issue as we are unable to uninstall it.
The checkmark is greyed out, and the *Uninstall* button (red cross) is disabled.
Why is that?
Some plugins have dependencies on other plugins, which in turn have dependencies on further plugins, creating a chain of dependencies.
When you put your mouse pointer over the red cross, you'll see a tooltip that indicates the parent plugin blocking the uninstallation:

image::/images/post-images/2023/06/19/2023-06-19-remove-outdated-plugins-while-using-docker/who-is-my-daddy.png[what is the parent plugin blocking the uninstallation?]

In this case, `popper` is a dependency for another plugin called `bootstrap4-api`.
Therefore, we need to remove `bootstrap4-api` first and then proceed with `popper`.

> Back to the [.line-through]#drawing board# search box, this time with `bootstrap4-api`.

Reviewer: ```suggestion
Back to the +++<del>+++drawing board+++</del>+++ search box, this time with `bootstrap4-api`.
```
In the deploy, it appears that the line-through is not rendering. I was searching around looking for an alternative that may work, and in my test it looks like this does the job.
<img width="1055" alt="Screenshot 2023-06-16 at 3 45 44 PM" src="https://github.com/jenkins-infra/jenkins.io/assets/99040580/65c6daa2-a2bb-4daa-a596-897a6abbaf12">

image::/images/post-images/2023/06/19/2023-06-19-remove-outdated-plugins-while-using-docker/bootstrap4-can-be-uninstalled.png[width=839,search for the parent plugin]

> This time, we can uninstall it by clicking on the red cross.

Reviewer: ```suggestion
This time, we can uninstall it by selecting the uninstall icon (red cross).
```
suggestion for selecting

> We will then have a warning message saying:

Reviewer: ```suggestion
We will then encounter a warning message saying:
```
updating seeing to encounter for accessibility

> You are about to uninstall the Bootstrap 4 API Plugin plugin.
This will remove the plugin binary from your $JENKINS_HOME, but it will leave the configuration files of the plugin untouched.

image::/images/post-images/2023/06/19/2023-06-19-remove-outdated-plugins-while-using-docker/remove-the-plugin-binary.png[width=839,remove the plugin binary]

> Really? We'll check that later. Click on **Yes** to proceed with the uninstallation.
> And we're back to the Installed plugins page.
> Let's give another chance to popper by searching for it again:

Reviewer: ```suggestion
Really?
We'll check that later. 
Select *Yes* to proceed with the uninstallation, and we're back to the Installed plugins page.
Let's give another chance to popper by searching for it again:
```
suggestion for formatting/sentence structure

image::/images/post-images/2023/06/19/2023-06-19-remove-outdated-plugins-while-using-docker/popper-can-be-uninstalled.png[width=839,popper can be uninstalled now]

> Same player, shoot again. Follow the same steps as before to uninstall `popper`.
> After successfully uninstalling popper, you may notice that the notification icon still displays a message.
> Furthermore, if we go back to the **Installed plugins** page, we'll see that `popper` is still there.

Reviewer: ```suggestion
Same player, shoot again.
Follow the same steps as before to uninstall `popper`.
After successfully uninstalling popper, you may notice that the notification icon still displays a message.
Furthermore, if we go back to the *Installed plugins* page, we'll see that `popper` is still there.
```
small suggestion for formatting

image::/images/post-images/2023/06/19/2023-06-19-remove-outdated-plugins-while-using-docker/pending-uninstallation.png[width=839,pending uninstallation]

> Why is this the case?
> We asked for uninstallation, but it didn't fully happen. Jenkins has to restart in order to complete the process.
> You can hit the `/safeRestart` endpoint to restart Jenkins safely and then click **Yes**.
> When you return, you will notice that the notification icon has disappeared, and the plugin is no longer listed on the Installed plugins page.

Reviewer: ```suggestion
Why is this the case?
We asked for uninstallation, but it didn't fully happen.
Jenkins has to restart in order to complete the process.
You can hit the `/safeRestart` endpoint to restart Jenkins safely and then select *Yes*.
When you return, you will notice that the notification icon has disappeared, and the plugin is no longer listed on the *Installed plugins* page.
```
suggestion to bold UI and adjust sentence structure/formatting

> ### Removing deprecated plugins in the docker context

Reviewer: ```suggestion
=== Removing deprecated plugins in the Docker context
```
small conversion for asciidoc and capitalization

However, depending on your Jenkins configuration, you may find that the deprecated plugins have somehow reappeared in your Jenkins instance, sometimes even with an older version.
How is this possible?
If your Jenkins container instance inherits from the Jenkins official container, it comes with a predefined set of plugins.
Most of the time, these plugins won't be enough for your specific use case.
You will need to install additional plugins.
When you do so, the new plugins will be installed in the `$JENKINS_HOME/plugins` directory, with a command such as:

[source,dockerfile]
----
COPY --chown=jenkins:jenkins ./plugins.txt /usr/share/jenkins/plugins.txt
RUN jenkins-plugin-cli --plugin-file=/usr/share/jenkins/plugins.txt
----

So...
Whenever you remove a deprecated plugin from the Jenkins UI, remember to also remove it from the Docker context.
Otherwise, it will be reinstalled when you rebuild the container.
In my case, I had to remove the following plugins from the `plugins.txt` file:

[source,dockerfile]
----
# See https://github.com/jenkinsci/docker#usage-1
ant:487.vd79d090d4ea_e
[...]
bootstrap4-api:4.6.0-3
[...]
popper-js:2.9.2-1
[...]
ws-cleanup:0.45
----

> Now you're safe for the next time you rebuild your Jenkins container.
> But what about your running container? Is it free of any reference to the deprecated plugins?
> Let's find out.
> 
> ### Removing deprecated plugins from the running container

Reviewer: ```suggestion
Now you're safe for the next time you rebuild your Jenkins container.
But what about your running container?
Is it free of any reference to the deprecated plugins?
Let's find out.

=== Removing deprecated plugins from the running container
```
small formatting suggestion

Here is an excerpt of my `docker-compose.yml` file:

[source,dockerfile]
----
#  docker compose up -d --build --force-recreate
services:
    jenkins:
        build: ./controller
        restart: always
        ports:
            - "8080:8080"
            - "50000:50000"
        volumes:
            - jenkins-data:/var/jenkins_home:rw
            - ./casc.d:/var/jenkins_home/casc.d/:ro
        environment:
            - CASC_JENKINS_CONFIG=/var/jenkins_home/casc.d/
[...]
volumes:
    jenkins-data:
----

The `jenkins-data` volume is mounted on the `/var/jenkins_home` directory of the container.
However, the `/usr/share/jenkins/plugins.txt` file, as we saw earlier in the `Dockerfile`, is not mounted on a shared volume.

I happen to have installed `bash` in my container, so I can run the following command to get a shell in the container (`jenkins` is the name of the service in the `docker-compose.yml` file):

[source,bash]
----
docker compose exec -it jenkins bash
----

> You could do the same with `sh` if `bash` was not installed in your Docker image.

Reviewer: ```suggestion
You can do the same with `sh` if `bash` was not installed in your Docker image.
```
small syntax suggestion

Now, let's search for the plugins definition file.
As we've seen in the Dockerfile, it's located in `/usr/share/jenkins/plugins.txt`:

[source,bash]
----
cd /usr/share/jenkins
cat plugins.txt |grep bootstrap4-api
bootstrap4-api:4.6.0-3
----

The reference to the deprecated plugin is still there.
Is that a problem?

> No. As the link:https://github.com/jenkinsci/docker#usage-1[documentation] says:

Reviewer: ```suggestion
No.
As the link:https://github.com/jenkinsci/docker#usage-1[documentation] says:
```
small suggestion for formatting

> When jenkins container starts, it will check JENKINS_HOME has this reference content, and copy them there if required. It will not override such files, so if you upgraded some plugins from UI they won't be reverted on next start.

So it's there, but it won't do any harm, it won't be used... unless we restart Jenkins.
Let's leave it there, until the next time we rebuild the container, as we have already cleaned up the `plugins.txt` file used by the Docker context earlier.

> Now what? Let's have a look at the `$JENKINS_HOME` directory.

Reviewer: ```suggestion
Now what?
Let's have a look at the `$JENKINS_HOME` directory.
```
small formatting suggestion

[source,bash]
----
cd $JENKINS_HOME
find . -name plugins.txt
----

> Nothing. We don't have a `plugins.txt` file in the `$JENKINS_HOME` directory.

Reviewer: ```suggestion
Nothing.
We don't have a `plugins.txt` file in the `$JENKINS_HOME` directory.
```
small formatting suggestion

I'm afraid we can.

[source,bash]
----
find . -name bootstrap4*
./plugins/bootstrap4-api
./plugins/bootstrap4-api/META-INF/maven/io.jenkins.plugins/bootstrap4-api
./plugins/bootstrap4-api/WEB-INF/lib/bootstrap4-api.jar
./plugins/bootstrap4-api.bak
./plugins/bootstrap4-api.jpi
./plugins/bootstrap4-api.jpi.version_from_image
./plugins/bootstrap4-api.jpi.pinned
----

There are still some traces of the `bootstrap4-api` deprecated plugin in the `$JENKINS_HOME/plugins` directory.
What about the `popper-js` plugin?
It's there too.
It may explain why despite having removed the deprecated plugins from the Jenkins UI, they were still there when we restart the container.
Let's remove them for real this time:

[source,bash]
----
rm -rf ./plugins/bootstrap4-api*
rm -rf ./plugins/popper*
----

We can now safely exit the container and restart it from the UI by accessing the `/safeRestart` endpoint.
Once we return, we should verify that the deprecated plugins are no longer present.

image::/images/post-images/2023/06/19/2023-06-19-remove-outdated-plugins-while-using-docker/the-following-plugins-are-deprecated.png[width=839,deprecated plugins warning]

> Oh no! It seems like the deprecated plugins have reappeared in the running container.

Reviewer: ```suggestion
Oh no!
It seems like the deprecated plugins have reappeared in the running container.
```
small formatting suggestion

How did that happen?
It's because we only restarted the container without rebuilding it.
The configuration still references the deprecated plugins.

Simply restarting the container repeatedly won't resolve the issue.
We need to rebuild the image after removing the deprecated plugins from the Docker context.
Then, we can recreate the container and remove the deprecated plugins from the running container using the UI.

As a Jenkins admin, it's important to go with the flow and avoid swimming upstream like a salmon.
By following the proper steps, we can address this issue effectively.

> Want to try it by yourself? Just follow the steps of the *TL;DR* section.

Reviewer: ```suggestion
Want to try it by yourself? 
Just follow the steps of the *TL;DR* section.
```
small formatting suggestion
